tap_eip712_message/src/lib.rs

```diff
@@ -0,0 +1,120 @@
+//! let signer = signed_message.recover_signer(&domain_separator).unwrap();
+//!
+//! assert_eq!(signer, wallet_address);
+//! ```
+//!
+
+use alloy::{
+    dyn_abi::Eip712Domain,
+    primitives::{Address, PrimitiveSignature as Signature},
+    signers::{local::PrivateKeySigner, SignerSync},
+    sol_types::SolStruct,
+};
+use serde::{Deserialize, Serialize};
+
+#[derive(thiserror::Error, Debug)]
+pub enum Eip712Error {
+    /// `alloy` wallet error
+    #[error(transparent)]
+    WalletError(#[from] alloy::signers::Error),
+
+    /// `alloy` wallet error
+    #[error(transparent)]
+    SignatureError(#[from] alloy::primitives::SignatureError),
+}
+
+/// EIP712 signed message
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq)]
+pub struct EIP712SignedMessage<M: SolStruct> {
+    /// Message to be signed
+    pub message: M,
+    /// ECDSA Signature of eip712 hash of message
+    pub signature: Signature,
+}
+
+#[derive(Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
+pub struct SignatureBytes([u8; 65]);
+
+pub trait SignatureBytesExt {
+    fn get_signature_bytes(&self) -> SignatureBytes;
+}
+
+impl SignatureBytesExt for Signature {
+    fn get_signature_bytes(&self) -> SignatureBytes {
+        SignatureBytes(self.as_bytes())
+    }
+}
+
+/// Unique identifier for a message
+///
+/// This is equal to the hash of the contents of a message, excluding the signature.
+/// This means that two receipts signed by two different signers will have the same id.
+///
+///
+/// This cannot be used as a unique identifier for a message, but can be used as a key
+/// for a hashmap where the value is the message.
+#[derive(Debug, Eq, PartialEq, Hash)]
+pub struct MessageId(pub [u8; 32]);
+
+impl<M: SolStruct> EIP712SignedMessage<M> {
+    /// Creates a signed message with signed EIP712 hash of `message` using `signing_wallet`
+    pub fn new(
+        domain_separator: &Eip712Domain,
+        message: M,
+        signing_wallet: &PrivateKeySigner,
+    ) -> Result<Self, Eip712Error> {
+        let recovery_message_hash = message.eip712_signing_hash(domain_separator);
+
+        let signature = signing_wallet.sign_hash_sync(&recovery_message_hash)?;
+
+        Ok(Self { message, signature })
+    }
+
+    /// Recovers and returns the signer of the message from the signature.
+    pub fn recover_signer(&self, domain_separator: &Eip712Domain) -> Result<Address, Eip712Error> {
+        let recovery_message_hash = self.message.eip712_signing_hash(domain_separator);
+        let recovered_address = self
+            .signature
+            .recover_address_from_prehash(&recovery_message_hash)?;
+        Ok(recovered_address)
+    }
+
+    /// Checks that receipts signature is valid for given verifying key, returns `Ok` if it is valid.
+    ///
+    /// # Errors
+    ///
+    /// Returns [`crate::Error::SignatureError`] if the recovered address from the
+    /// signature is not equal to `expected_address`
+    ///
+    pub fn verify(
```

Comment:
```suggestion
    /// Checks that receipts signature is valid for given verifying key, returns `Ok(true)` if it is valid.
    ///
    /// # Errors
    ///
    /// Returns [`crate::Error::SignatureError`] if the recovered address from the
    /// signature is not equal to `expected_address`
    ///
    pub fn verify(
```

And then error from `recover_signer`

Reply:
Done